src/typechecker/FStar.TypeChecker.TcTerm.fst

```diff
@@ -3114,19 +3114,19 @@ and tc_pat env (pat_t:typ) (p0:pat) :
                 Some (S.withinfo (Pat_wild x) p.p, false))
               p.p
           in
-          let p = { p with v=Pat_cons(constructor_fv, sub_pats) } in
+          let p = { p with v=Pat_cons(constructor_fv, us_opt, sub_pats) } in
           let p = PatternUtils.elaborate_pat env p in
           check_nested_pattern env p t
 
-        | Pat_cons(fv, sub_pats) ->
+        | Pat_cons(fv, us_opt, sub_pats) ->
           let simple_pat =
             let simple_sub_pats =
                 List.map (fun (p, b) ->
                     match p.v with
                     | Pat_dot_term _ -> p, b
                     | _ -> S.withinfo (Pat_var (S.new_bv (Some p.p) S.tun)) p.p, b)
                 sub_pats in
-            {p with v = Pat_cons (fv, simple_sub_pats)}
+            {p with v = Pat_cons (fv, None, simple_sub_pats)}
```

Comment:
The `None` here is temporary, to be filled-in later? Could use `us_opt` from line 3121 also?

Reply:
I changed this so that if we have a `Pat_cons(_, Some us, _)` then tc_eqn uses the provided instantiation, rather than recomputing it.

The overall effect is that in phase1, we start with a `Pat_cons(_, None, _)` which then gets filled to `Pat_cons(_, Some us, _)` and phase2 checks that instantiation rather than recomputing it